Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp

```diff
@@ -177,6 +177,168 @@ bool TerrainSystem::InWorldBounds(float x, float y) const
+
+    size_t indexStepSize = 1;
+    if (sampler == AzFramework::Terrain::TerrainDataRequests::Sampler::BILINEAR)
+    {
+        // outPositions holds the iterators to results of the bulk queries.
+        // In the case of the bilinear sampler, we'll be making 4 queries per
+        // input position.
+        outPositions.resize(inPositions.size() * 4);
+        outTerrainExists.resize(inPositions.size() * 4);
+        indexStepSize = 4;
+    }
+    else
+    {
+        outPositions.resize(inPositions.size());
+        outTerrainExists.resize(inPositions.size());
+    }
+
+    GenerateQueryPositions(inPositions, outPositions, sampler, indexStepSize);
+
+    AZ::Aabb bounds;
+    AZ::EntityId prevAreaId = FindBestAreaEntityAtPosition(outPositions[0].GetX(), outPositions[0].GetY(), bounds);
+    
+    // We use a sliding window here and update the window end for each
+    // position that falls in the same area as the previous positions. This consumes lesser memory
+    // than sorting the points into separate lists and handling putting them back together.
+    // This may be sub optimal if the points are randomly distributed in the list as opposed
+    // to points in the same area id being close to each other.
+    size_t windowStart = 0;
+    size_t windowEnd = 0;
+    const size_t numPositions = outPositions.size();
+    for(int i = 1; i < numPositions; i++)
+    {
+        AZ::EntityId areaId = FindBestAreaEntityAtPosition(outPositions[i].GetX(), outPositions[i].GetY(), bounds);
+        bool queryHeights = false;
+        if (areaId == prevAreaId)
+        {
+            // Update window end to current position.
+            // If it's the last position, submit the query.
+            windowEnd = i;
+            if (windowEnd == numPositions - 1)
+            {
+                queryHeights = true;
+            }
+        }
+        else
+        {
+            queryHeights = true;
+        }
+
+        if (queryHeights)
+        {
+            size_t spanLength = (windowEnd - windowStart) + 1;
+            Terrain::TerrainAreaHeightRequestBus::Event(prevAreaId,
+                &Terrain::TerrainAreaHeightRequestBus::Events::GetHeights,
+                AZStd::span<AZ::Vector3>(outPositions.begin() + windowStart, spanLength), AZStd::span<bool>(outTerrainExists.begin() + windowStart, spanLength));
+            
+            // Reset the window to start at the current position.
+            // Set the new area id on which to run the next query.
+            windowStart = windowEnd = i;
+            prevAreaId = areaId;
+        }
+    }
+
+    // Compute/store the final result
+    for (size_t i = 0, iteratorIndex = 0; i < inPositions.size(); i++)
+    {
+        if (!InWorldBounds(inPositions[i].GetX(), inPositions[i].GetY()))
+        {
+            // Not in world bounds. Store defaults and move on.
+            heights[i] = minHeight;
+            terrainExists[i] = false;
+            continue;
+        }
```

Comment:
This check is getting done twice for inPositions, once here and once in GenerateQueryPositions. It's probably fast enough that anything else isn't worth it, but pointing it out in case you've got any clever ideas for avoiding the double-check.

Reply:
I fixed an edge case I missed in this [commit](https://github.com/o3de/o3de/pull/7357/commits/22e6ff72573cee957fced8edbd54fd678cfa7722). Also added a comment about it. I think we can get rid of the multiple world bounds check and instead handle it the way I do in that commit. I haven't removed the checks yet but with that commit, I think those checks are now redundant. I'll run the unit tests after getting rid of them to confirm.

Reply:
There may be something even simpler. I can get rid of all the worlds bounds checks including the one I added for the edge case. If I initialize the z component of all the positions to min height when we're generating the positions, then the calls to GetHeights for the positions that are outside of any registered areas(this covers the outside world bounds case as well) won't get routed anywhere since the area id is invalid. This will make the code much simpler to read. With a lot of registered areas, it may be cheaper to do world bounds check. But we have to call FindClosestAreaForPosition() on all valid positions anyway. Only very few positions that are really outside bounds would have been cheaper with a bounds check. We may actually be doing much better if we remove the world bounds check and only use the FindAreaId function to filter. Because for valid positions, the world bounds check ends up being redundant.